File: ui/src/types/blueprint.ts
export interface Blueprint {
    uuid: string;
    standardName: string;
    type: string;
    description: string;
    functionalities?: Functionality[];
    libraries?: Library[];
}

export interface Library {
    uuid: string;
    standardName: string;
    version: string;
    description: string;
    repositoryURL: string;
    namespace: string;
    exposedFunctionalities?: LibraryFunctionality[];
}

export interface LibraryFunctionality {
    uuid: string;
    name: string;
    type: string;
    description: string;
}

export interface Functionality {
    uuid: string;
    category: string;
    type: string;
    provider: string;
    implementsGenerics: boolean;
    filePathsCSV: string;
    operations?: FunctionalOperation[];
}

export interface FunctionalOperation {
    uuid: string;
    name: string;
    description: string;
}

File: src/domain/model/blueprint.go
package model

type Blueprint struct {
	BaseModel
	StandardName    string          `gorm:"size:255"`
	Type            string          `gorm:"size:100"`
	Description     string          `gorm:"size:1000"`
	Functionalities []Functionality `gorm:"foreignKey:BlueprintID;constraint:OnUpdate:CASCADE,OnDelete:CASCADE"`
	Libraries       []Library       `gorm:"many2many:blueprint_libraries"`
}

type Functionality struct {
	BaseModel
	Category           string `gorm:"size:100"`
	Type               string `gorm:"size:100"`
	Provider           string `gorm:"size:100"`
	ImplementsGenerics bool
	FilePathsCSV       string `gorm:"size:1000"`
	BlueprintID        uint
	Operations         []FunctionalOperation `gorm:"foreignKey:FunctionalityID;constraint:OnUpdate:CASCADE,OnDelete:CASCADE"`
}

type FunctionalOperation struct {
	BaseModel
	Name            string `gorm:"size:255"`
	Description     string `gorm:"size:1000"`
	FunctionalityID uint
}
